nav2_experimental_rl/turtlebot3_env.py

```diff
@@ -0,0 +1,148 @@
+
+        self.reset_simulation = node_.create_client(Empty, 'reset_simulation')
+        self.reset_world = node_.create_client(Empty, 'reset_world')
+        self.unpause_proxy = node_.create_client(Empty, 'unpause_physics')
+        self.pause_proxy = node_.create_client(Empty, 'pause_physics')
+        self.get_entity_state = node_.create_client(GetEntityState, 'get_entity_state')
+        self.set_entity_state = node_.create_client(SetEntityState, 'set_entity_state')
+
+        t = Thread(target=rclpy.spin, args=[node_])
+        t.start()
+
+    def get_reward(self):
+        reward = 0
+        if self.collision == True:
+          reward = -10
+          self.done = True
+          return reward, self.done
+        elif self.collision == False and self.act == 0:
+          if abs(min(self.states_input)) >= self.zero_div_tol:
+           reward = 0.08 - (1/(min(self.states_input)**2))*0.005
+          else:
+           reward = -10
+          if reward > 0:
+           self.bonous_reward += reward
+           reward = self.bonous_reward
+        else:
+          bonous_discount_factor = 0.6
+          self.bonous_reward *= bonous_discount_factor
+          if abs(min(self.states_input)) >= self.zero_div_tol:
+           reward = 0.02 - (1/min(self.states_input))*0.005
+          else:
+           reward = -10
+        return reward, self.done
+
+
+    def scan_callback(self, LaserScan):
+        self.laser_scan_range = []
+        self.range_min = LaserScan.range_min
+        range_max = LaserScan.range_max
+
+        for i in range(len(LaserScan.ranges)):
+            if LaserScan.ranges[i] == float('Inf'):
+               self.laser_scan_range.append(range_max)
+            elif LaserScan.ranges[i] < self.range_min:
+               self.laser_scan_range.append(self.range_min + self.collision_tol)
+            else :
+               self.laser_scan_range.append(LaserScan.ranges[i])
+        if self.check_collision():
+          self.collision = True
+          self.done = True
+        self.states_input = []
+        for i in range(8):
+          step = int(len(LaserScan.ranges)/8)
+          self.states_input.append(min(self.laser_scan_range[i*step:(i+1)*step], default=0))
+
+    def action_space(self):
+        return len(self.actions)
+
+    def step(self, action):
+        vel_cmd = Twist()
+        self.act = action
+        vel_cmd.linear.x = self.actions[action][0]
+        vel_cmd.angular.z = self.actions[action][1]
+        self.pub_cmd_vel.publish(vel_cmd)
+        vel_cmd.linear.x = 0.0
+        vel_cmd.angular.z = 0.0
+        get_reward = self.get_reward()
+        return self.states_input, get_reward[0], self.done
+    
+    def check_collision(self):
+        if  min(self.laser_scan_range) < self.range_min + self.collision_tol:
+         print("Near collision detected... " + str(min(self.laser_scan_range)))
+         return True
+        return False
+
+    def reset(self):
```

Comment:
I found in this case, you should wait for measurements before exiting. ie use the `wait_for_message` type primative, that way you dont start processing control efforts before you have new information from your new trial. 0-ing out will probably get rid of issues were stagnated data are being processed in a new run, but it doesn't represent the actual envrionment you're training the agent to react to

Reply:
It seems like in rclpy 'wait_for_message' is not available.  I used a boolean flag in the subscriber callback as a temporary workaround.